lib/SILOptimizer/PassManager/PassPipeline.cpp

```diff
@@ -758,6 +758,11 @@ static void addLowLevelPassPipeline(SILPassPipelinePlan &P) {
 
   addFunctionPasses(P, OptimizationLevelKind::LowLevel);
 
+  // The NamedReturnValueOptimization shouldn't be done before serialization,
+  // because it might prevent predictable memory optimizations in the caller
+  // after inlining.
```

Comment:
Thanks for commenting this pipeline restriction. Pipeline comments are helpful, but I think we should also have a convention that pass dependencies are specified in the file that defines the pass. That's where people will see it when they are investigating the pass, and the information won't be lost as easily. I regularly see the pipeline comments get messed up as people hack on it. In the pass file, you can describe the dependency in more detail with examples. In this case, I'm not sure why NRVO interferes with predictable mem opts, but deferring it until after serialization seems reasonable. It doesn't seem needed for canonicalization and might not be useful if inlining happens.

Reply:
👍I added a comment